Document that SystemTime does not count leap seconds

ijackson opened the pull request:
Fixes #77994

This may not be entirely uncontroversial.  I know that @Ekleog is going to disagree.  However, in support of this docs change:

 This documents the current behaviour.  The alternative would be to plan to *change* the behaviour.

There are many programs which need to get a POSIX time (a `time_t`).  Right now, `duration_since(UNIX_EPOCH)` is the only facility in std that does that.  So, that is what programs use.  Changing the behaviour would  break[1] all of those programs.  We would need to define a new API that can be used to get a POSIX time, and get everyone to use it.  This seems highly unpalatable.

And, even if we wanted to do that, time with leap seconds is a lot less easy to work with.  We would need to arrange to have a leap seconds table available to `std` somehow, and make sure that it was kept up to date.  Currently we don't offer to do that for timezone data, which has similar needs.  There are other complications.  So it seems it would be awkwarrd to *implement* a facility that provides time including leap seconds, and the resulting value would be hard for applications to work with.

Therefore, I think it's clear that we don't want to plan to ever change `SystemTime`.  We should plan to keep it the way it is.  Providing TAI (for example) should be left to external crates, or additional APIs we may add in the future.

For more discussion see #77994 and in particular @fanf2's https://github.com/rust-lang/rust/issues/77994#issuecomment-1409448174

[1]  Of course, by "break" we really only mean *future* breakage in the case where there is, in fact, ever another leap second.  There may well not be: they are in the process of being abolished (although this is of course being contested).  But if we decide that `SystemTime::now().duraton_since(UNIX_EPOCH)` counts leap seconds, it would start to return `Durations`s that are 27s different to the current answers.   That's clearly unacceptable.  And we can hardly change `UNIX_EPOCH` by 27s.

ijackson committed:
Add note about clone-and-hack to the two UNIX_EPOCH definitions

I edited one of these and looked at the formatted docs for the other.
This confused me for a while; I suspected a build system bug.

I don't see an easy and neat way to unify these.  So let's just
document it instead.

ijackson committed:
Document that SystemTime does not count leap seconds

Fixes #77994

Ekleog commented:
My opposition is more of a principled opposition: I think we should eventually deprecate `duration_since` in favor of some other name that makes it clear that the returned number is not an actual duration.

However, I'm still all an favor of fixing the documentation so that at least the current behavior is well-documented :)

I'm just thinking this PR is missing the place that is actually most subject to issues, which is the `duration_since` method; as the following code might have very surprising behavior if presented with a leap second even on an otherwise well-behaved system… and it's actually the code example that's currently being given to `SystemTime::duration_since`: (including, I'd guess, probably panicking on some OSes, though hopefully none of the main ones)
```rust
let start = SystemTime::now();
do_stuff();
println!("Time it took to run do_stuff: {:?}", SystemTime::now().duration_since(start).unwrap())
```

thomcc commented:
Our behavior here is clearly an API decision, that requires some amount of discussion, so I'm nominating this.

m-ou-se commented:
We discussed this in a recent libs api meeting. We came to the same conclusion as above: the only way to convert a SystemTime to a date/time currently is by subtracting UNIX_EPOCH from it. If the resulting Duration would include leap seconds, an accurate conversion wouldn't be possible. (Even if we were to ship a database of leap seconds; future leap seconds aren't even known yet.)

So, SystemTime should not include leap seconds, and we should document that guarantee. Any other option would make SystemTime mostly unusable and incompatible with how it's already used in many places today.

ijackson commented:
> So, SystemTime should not include leap seconds, and we should document that guarantee. Any other option would make SystemTime mostly unusable and incompatible with how it's already used in many places today.

Cool, thanks.  That's what I'm intending with the text in this MR.  @thomcc, does that text look good to you?

thomcc commented on library/std/src/time.rs:
```diff
@@ -461,13 +469,22 @@ impl fmt::Debug for Instant {
+    /// returns a POSIX `time_t` (as a `u64`):
+    /// the number of non-leap seconds since the start of 1970 UTC.
+    /// This is the same time representation as used in many Internet protocols,
+    /// for example: JWT, CBOR, TOTP, certificate transparency and DNS TSIG/DNSSEC.
```
I think enumerating these probably is over the line for too much documentation, since it might lead people to file PRs adding other protocols. I also would prefer saying "This is $description, same as $POSIX_equivalent" rather than the other way around.

So my suggesion would be for this (and the other) comment to be more like:

> `duration_since(UNIX_EPOCH).unwrap().as_secs()` returns the number of non-leap seconds since the start of 1970 UTC, as a u64. This is the same time representation as used in POSIX `time_t`, as well as a wide variety of internet protocols.

(Note: I'm not tied to precise wording except for the bits I mentioned above). Thoughts?

ijackson replied:
I'm OK with these suggestions and have adopted them.  Thanks.

thomcc reviewed (commented):
Sorry for the delay, I had some family issues that caused me to fall behind on reviews.

A bit bikesheddy, but I would prefer these have their wording tweaked.

thomcc commented:
@rustbot author

ijackson committed:
Adjust the duration_since(UNIX_EPOCH) docs

 * Make the description primary, not the definition in terms of time_t
 * Remove the list of Internet protocols

As per
  https://github.com/rust-lang/rust/pull/109660#pullrequestreview-1414613118

ijackson commented:
@rustbot reviewer

thomcc commented:
@bors r+

bors commented:
:pushpin: Commit c4bc16c5d60ccaebbd975ce5e7bcde661045951c has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc c4bc16c5d60ccaebbd975ce5e7bcde661045951c -->
<!-- homu: {"type":"Approved","sha":"c4bc16c5d60ccaebbd975ce5e7bcde661045951c","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

thomcc commented:
@bors rollup
